test/integration/scheduler/serving/main_test.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduler
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"testing"
+
+	"k8s.io/klog/v2/ktesting"
+	kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	kubeschedulertesting "k8s.io/kubernetes/cmd/kube-scheduler/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+func TestMain(m *testing.M) {
+	framework.EtcdMain(m.Run)
+}
+
+func TestHealthEndpoints(t *testing.T) {
+	server, configStr, _, err := startTestAPIServer(t)
+	if err != nil {
+		t.Fatalf("Failed to start kube-apiserver server: %v", err)
+	}
+	defer server.TearDownFn()
+
+	apiserverConfig, err := os.CreateTemp("", "kubeconfig")
+	if err != nil {
+		t.Fatalf("Failed to create config file: %v", err)
+	}
+	defer func() {
+		_ = os.Remove(apiserverConfig.Name())
+	}()
+	if _, err = apiserverConfig.WriteString(configStr); err != nil {
+		t.Fatalf("Failed to write config file: %v", err)
+	}
+
+	brokenConfigStr := strings.ReplaceAll(configStr, "127.0.0.1", "127.0.0.2")
+	brokenConfig, err := os.CreateTemp("", "kubeconfig")
+	if err != nil {
+		t.Fatalf("Failed to create config file: %v", err)
+	}
+	if _, err := brokenConfig.WriteString(brokenConfigStr); err != nil {
+		t.Fatalf("Failed to write config file: %v", err)
+	}
+	defer func() {
+		_ = os.Remove(brokenConfig.Name())
+	}()
+
+	tests := []struct {
+		name             string
+		path             string
+		useBrokenConfig  bool
+		wantResponseCode int
+	}{
+		{
+			"/healthz",
+			"/healthz",
+			false,
+			http.StatusOK,
+		},
+		{
+			"/livez",
+			"/livez",
+			false,
+			http.StatusOK,
+		},
+		{
+			"/livez with ping check",
+			"/livez/ping",
+			false,
+			http.StatusOK,
+		},
+		{
+			"/readyz",
```

Comment:
Honestly, I'm not convinced that the tests here covered all the logics, I think we need more like:
- livez failed for unavailable APIServer
- readyz failed for handler unsynced
- readyz failed for follower APIServer
- readyz failed for shutting down APIServer
- readyz succeeded for leader synced with handlers.

But have no idea whether this is easy to mock. We may need more faked healthChecker.

Reply:
These are valid points. However, there is no easy way to write those tests. OTOH, I don't think mocking would buy much for us in this case because what we truly want is to see how `/livez and /readyz` really behave in mentioned failed cases whereas mocking changes the healthcheck handlers directly without setting up the failure cases.

It's feasible to pass faked checkers to scheduler when setting scheduler up but some code is needed in scheduler to handle those test cases when faked checkers are provided. which makes me feel the actual testing is for the code that sets up the faked checkers. 

I also checked apiserver code related to healthcheck endpoints but there are only tests for successful cases so no prior work can be followed.

Reply:
Valid points to me, it's hard to mock like `waiting for handlers sync` scenarios unless we have a plenty of API objects to handle or inject some waiting logic to the scheduler handlers, so I'll put the approval tag here, and see will others have different opinions, btw, API review should still be weighted in.
/approve  cc @kubernetes/sig-scheduling-approvers We'll have two more health checks /readyz & /livez for scheduler.